Data access API for rcParams (#24730)
matplotlib/matplotlib: doc/api/next_api_changes/deprecations/24730-TH.rst

rcParams type
~~~~~~~~~~~~~
Relying on ``rcParams`` being a ``dict`` subclass is deprecated.

Nothing will change for regular users because ``rcParams`` will continue to
be dict-like (technically fulfill the ``MutableMapping`` interface).

The `.RcParams` class does validation checking on calls to

> `.RcParams.__get_item__` and `.RcParams.__set_item__`.  However, there are rare

Reviewer: Don't think there's any need to link these? Also, the name is wrong.

```suggestion
``.RcParams.__getitem__`` and ``.RcParams.__setitem__``.  However, there are rare
```

> However, there are some rare known cases that use
> ``dict.__getitem__(rcParams, key)`` and
> ``dict.__setitem__(rcParams, key, val)`` for direct data access that
> circumvents any validation logic. While this is technically public, it
> depends on the implementation detail that ``rcParams`` is currently a ``dict``
> subclass.
> 

Reviewer 2: OK, part of my confusion about this was that I didn't quite understand that this  was calling the `super.__getitem__`.  Suggested rewording for this to be clearer:

```suggestion
The `.RcParams` class does validation checking on calls to `.RcParams.__get_item__` and 
`.RcParams.__set_item__`.  However, there are rare cases where we want to circumvents the 
validation logic and directly access the underlying data values.   Previously, this could be 
accomplished via  a call to the parent methods  ``dict.__getitem__(rcParams, key)`` and ``dict.__setitem__(rcParams, key, val)``. 

```

Matplotlib 3.7 introduces ``rcParams._set(key, val)`` and
``rcParams._get(key)`` as a replacement to calling the parent methods. They are
intentionally marked private because there is little need for external users

> to use them. However, if you have a compelling reason for direct `.RcParam`

Reviewer: ```suggestion
to use them. However, if you have a compelling reason for direct `.RcParams`
```

access, please switch from the dict functions to the new ``_get()`` and
``_set()``. Even though marked private, we guarantee API stability for these

> methods, i.e. they are subject to Matplotlib's API and deprecation policy.

Reviewer 3: ```suggestion
methods, i.e., they are subject to Matplotlib's API and deprecation policy.
```

> Matplotlib 3.7 introduces ``rcParams._set(key, val)`` and
> ``rcParams._get(key)`` as a replacements for these cases. They are
> intentionally marked private because there is very little need for external
> users to use them. However, if you have a compelling reason for direct data
> access, please switch from the dict functions to the new ``_get()`` and
> ``_set()``. Even though marked private, we guarantee API stability for these
> methods, i.e., they are subject to Matplotlib's API and deprecation policy.

Reviewer 2: ```suggestion
Matplotlib 3.7 introduces ``rcParams._set(key, val)`` and
``rcParams._get(key)`` as a replacement to calling the parent methods. They are
intentionally marked private because there is little need for external
users to use them. However, if you have a compelling reason for direct `.RcParam` data
access, please switch from the dict functions to the new ``_get()`` and
``_set()``. Even though marked private, we guarantee API stability for these
methods and they are subject to Matplotlib's API and deprecation policy.
```

Please notify the Matplotlib developers if you rely on ``rcParams`` being a

> dict subclass in any other way,for which there is no migration path yet.

Reviewer 3: ```suggestion
dict subclass in any other way, for which there is no migration path yet.
```
